crates/services/src/seqlock.rs

```diff
@@ -0,0 +1,165 @@
+//! A simple implementation of a sequential lock.
+//! More details: <https://docs.kernel.org/locking/seqlock.html>
+
+use std::{
+    cell::UnsafeCell,
+    panic::UnwindSafe,
+    sync::atomic::{
+        AtomicU64,
+        Ordering,
+    },
+};
+
+/// A simple implementation of a sequential lock.
+/// some usage of unsafe, T must be Copy
+#[derive(Debug)]
+pub struct SeqLock<T: Copy> {
+    sequence: AtomicU64,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl<T: Send + Copy> Sync for SeqLock<T> {}
+
+/// The writer handle for the `SeqLock`.
+/// Only one writer exists for a `SeqLock`.
+/// There is no Clone bound since we want to enforce only one writer.
+#[derive(Debug)]
+pub struct SeqLockWriter<T: Copy> {
+    lock: std::sync::Arc<SeqLock<T>>,
+}
+
+impl<T: Copy> SeqLockWriter<T> {
+    /// Modifies the data within the lock.
+    pub fn write<F>(&self, f: F)
+    where
+        F: FnOnce(&mut T) + UnwindSafe,
+    {
+        let lock = &self.lock;
+
+        // Indicate that a write operation is starting.
+        lock.sequence.fetch_add(1, Ordering::Release);
+
+        // attempt to perform the write, and catch any panics
+        // safety: panics are caught and resumed
+        let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| unsafe {
+            let data = &mut *lock.data.get();
+            f(data);
+        }));
+
+        // Indicate that the write operation has finished.
+        lock.sequence.fetch_add(1, Ordering::Release);
+
+        // resume unwinding if there was an error
+        if let Err(e) = result {
+            std::panic::resume_unwind(e);
+        }
+    }
+}
+
+/// The reader handle for the `SeqLock`.
+/// Multiple readers can be created for a `SeqLock`.
+#[derive(Clone, Debug)]
+pub struct SeqLockReader<T: Copy> {
+    lock: std::sync::Arc<SeqLock<T>>,
+}
+
+impl<T: Copy> SeqLockReader<T> {
+    /// Reads the data within the lock.
+    pub fn read(&self) -> T {
+        let lock = &self.lock;
+
+        loop {
+            // check starting guard
+            let start = lock.sequence.load(Ordering::Acquire);
+
+            // if odd, write in progress
+            if start % 2 != 0 {
+                continue;
+            }
+
+            let data = unsafe { *lock.data.get() };
+
+            // check starting/ending guard
+            let end = lock.sequence.load(Ordering::Acquire);
+
+            // if value changed, retry
+            if start == end && start % 2 == 0 {
+                return data;
+            }
+        }
+    }
+}
+
+impl<T: Copy> SeqLock<T> {
+    /// Creates a new `SeqLock` and returns a writer and a reader handle.
+    /// Optimized for occasional writes and frequent reads
+    ///  !!WARNING!!
+    /// ONLY USE IF ALL THE BELOW CRITERIA ARE MET
+    ///  1. Internal data < 64 bytes
+    ///  2. ONLY 1 writer
```

Comment:
This is already ensured by the type, so having it inside `!!WARNING!!` section seems excessive

Reply:
addressed in 51d079ca61